src/coordination/coordinator_instance.cpp

```diff
@@ -166,8 +102,42 @@ auto CoordinatorInstance::TryFailover() -> void {
     return;
   }
 
-  // TODO: Smarter choice
-  auto new_main = ranges::begin(alive_replicas);
+  std::string most_up_to_date_instance;
+  std::optional<uint64_t> latest_commit_timestamp;
+  std::optional<std::string> latest_epoch;
+  {
+    // for each DB in instance we get one DatabaseHistory
+    using DatabaseHistories = replication_coordination_glue::DatabaseHistories;
+    std::vector<std::pair<std::string, DatabaseHistories>> instance_database_histories;
+
+    bool success{true};
+    std::for_each(alive_replicas.begin(), alive_replicas.end(),
+                  [&success, &instance_database_histories](ReplicationInstance &replica) {
+                    if (!success) {
+                      return;
+                    }
+                    auto res = replica.GetClient().SendGetInstanceTimestampsRpc();
+                    if (res.HasError()) {
+                      spdlog::error("Could get per db history data for instance {}", replica.InstanceName());
+                      success = false;
+                      return;
+                    }
+                    instance_database_histories.emplace_back(replica.InstanceName(), std::move(res.GetValue()));
+                  });
```

Comment:
I understand the logic but we can write this nicer:

```cpp
auto maybe_db_histories = std::ranges::transform(alive_replicas, [](ReplicationInstance &replica) {
  return replica.GetClient().SendGetInstanceTimestampsRpc();
}) | ranges::to<replication_coordination_glue::DatabaseHistories>();
if (std::ranges::any_of(maybe_db_histories, [](auto &maybe_history) {
  return maybe_history.HasError();
)) {
spdlog::error("Aborting failover as at least one instance didn't provide per database history.");
return;
}

auto db_histories = std::ranges::transform(maybe_histories, [](auto &maybe_history) {
  return maybe_history.GetValue();
} | ranges::to<std::vector<std::pair<...>>();





```

Reply:
I didn't write 100% correct syntax but you will know what I mean

Reply:
Aren't we sending this way more RPCs then needed? If it fails on the first message, we will send still 2 or 3 RPCs. I am not sure if we should optimize for code readability here or resource protection

Reply:
TODO: write in more readable way with transform, transform_if

Reply:
I would argue this is not more readable, so will not do it:
```
// for each DB in instance we get one DatabaseHistory
    using DatabaseHistories = replication_coordination_glue::DatabaseHistories;

    using GetInstanceRetType = decltype(std::declval<CoordinatorClient>().SendGetInstanceTimestampsRpc());
    using MaybeInstanceValuePairType = std::pair<std::string, GetInstanceRetType>;
    auto maybe_db_histories = ranges::views::transform(alive_replicas, [](ReplicationInstance &replica) {
                                return std::make_pair<std::string, GetInstanceRetType>(replica.InstanceName(), replica.GetClient().SendGetInstanceTimestampsRpc());
                              }) | ranges::to<std::vector<MaybeInstanceValuePairType>>();


    if (std::ranges::any_of(maybe_db_histories, [](const auto &maybe_history) {
          return std::get<1>(maybe_history).HasError();
        })) {
      spdlog::error("Aborting failover as at least one instance didn't provide per database history.");
      return;
    }

    using InstanceValuePairType = std::pair<std::string, DatabaseHistories>;
    std::vector<InstanceValuePairType> instance_database_histories;

    std::ranges::transform(maybe_db_histories, std::back_inserter(instance_database_histories), [](auto &value){
      return std::pair<std::string, DatabaseHistories>(std::get<0>(value), std::get<1>(value).GetValue()); 
    });
```